Streaming post counts

jayzes opened the pull request:
This implements a basic setup for streaming post/author impression counts via Spark/Kinesis.

## Open questions/things to investigate:

- [x] How do we deploy this? EMR looks promising (and potentially easy), though @rtyer mentioned that he's had difficulty setting that up previously

Looking into EMR with Cloudformation. Seems like a reasonable solve, absent deployment issues with the JARs.

- [x] How do we deal with updates/configuration changes without losing data? The resumable checkpointing behavior seems to solve a lot of this, but you still can't change the configuration or class names after the fact, which seems like a problem.

    We'll have to implement our own checkpointing to S3 in some kind of file format.

- [x] How do we get data into nearline storage for serving via the primary API? I'm thinking that Redis (via redis-spark) is a good option, though we could do Postgres if we wanted to.

    Yes, redis-spark should be fine. Don't want to do snapshots from it though since Redis `scan`s are not parallelizable.

- [x] How do we test this in an automated way?

    `scala-test` and `spark-testing-base` are just peachy.

- [ ] How do we monitor to make sure that throughput is appropriate? How do we know if the job starts swapping out?

    Looks like we can get metrics from the KCL library (http://docs.aws.amazon.com/streams/latest/dev/monitoring-with-kcl.html#kcl-metrics-per-app) as well as EMR itself (http://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/emr-metricscollected.html). We can also hook up Spark's `log4j` and metrics setups to Cloudwatch Logs and/or Cloudwatch itself.

- [x] Do we need to worry about reposts behaving differently than original posts?

    Good question. Two events are sent for each repost, so I think we should be good counting naively.

- [ ] Should we change this to use the new [structured streaming](http://spark.apache.org/docs/latest/structured-streaming-programming-guide.html) constructs? Those seem to be more stable in 2.1, [which is now out](https://spark.apache.org/news/spark-2-1-0-released.html)

## Deployment checklist:

- [ ] Disarticulate mothership count incrementing from domain event creation. Set up a new Redis instance for pulling counts.
- [ ] Stop count incrementing from the mothership and record the time of the last incremented count. Kinesis events should keep flowing. Record the last increment timestamp to pass to the cold start process.
- [ ] Manually create an initial snapshot file from Redis with the initial post counts. Save for posterity and/or backup. Drop any events created before the snapshot was created.
- [ ] Cold-start this process and begin aggregating counts.
- [ ] Remove old counts from the original Redis instance.

https://github.com/pdeyhim/spark-emr/blob/master/core/src/main/scala/org/apache/spark/metrics/sink/CloudwatchSink.scala
https://blog.cloudera.com/blog/2015/03/how-to-tune-your-apache-spark-jobs-part-2/
https://jaceklaskowski.gitbooks.io/mastering-apache-spark/content/spark-executor.html
https://www.slideshare.net/AmazonWebServices/best-practices-for-using-apache-spark-on-aws
https://vanwilgenburg.wordpress.com/2015/02/15/spark-tuning-guide/
http://mkuthan.github.io/blog/2016/09/30/spark-streaming-on-yarn/
https://www.slideshare.net/SparkSummit/recipes-for-running-spark-streaming-apploications-in-production-tathagata-daspptx

Commit:
Update ElloRecommend for Spark 2.0

Commit:
Aggregate live counts by post and author

http://spark.apache.org/docs/latest/streaming-kinesis-integration.html

https://docs.cloud.databricks.com/docs/latest/databricks_guide/07%20Spark%20Streaming/12%20Global%20Aggregations%20-%20mapWithState.html

https://docs.cloud.databricks.com/docs/latest/databricks_guide/07%20Spark%20Streaming/11%20Global%20Aggregations%20-%20updateStateByKey.html

Commit:
Update Java -> Scala seq usage to be more idiomatic

steam commented on src/main/scala/ElloStreamingCount.scala:
```diff
@@ -0,0 +1,166 @@
+    if (args.length != 3) {
+      System.err.println(
+        """
+        |Usage: ElloStreamingImpressionCount <app-name> <stream-name> <endpoint-url> <region-name>
```
Such useful output 🤘

steam commented on src/main/scala/ElloStreamingCount.scala:
```diff
@@ -0,0 +1,166 @@
+    require(credentials != null,
+      "No AWS credentials found. Please specify credentials using one of the methods specified " +
+      "in http://docs.aws.amazon.com/AWSSdkDocsJava/latest/DeveloperGuide/credentials.html")
+      val kinesisClient = new AmazonKinesisClient(credentials)
```
Line 60 starts a column of strange indentation. Should it be aligned with line 57?

jayzes replied:
Oh good call - I was thinking that was part of a new block scope but it's just a weird artifact.

steam commented on src/main/scala/ElloStreamingCount.scala:
```diff
@@ -0,0 +1,166 @@
+
+      // Setup the SparkConfig and StreamingContext
+      val sparkConfig = new SparkConf().setAppName("ElloStreamingImpressionCounts")
+      val ssc = new StreamingContext(sparkConfig, batchInterval)
```
Maybe a bit nit-picky but I'd name this `streamingContext` to provide more clarity.

jayzes replied:
Good call. This is going to get refactored a bit anyway to allow resumption of streaming after a shutdown - I'll rename it with that.

steam reviewed (commented):
I don't have much to contribute on the scala spark side since I am unfamiliar with it 😄 but it looks promising. 👍 

How are Scala unit tests handled? This is a WIP I know so it's not complete.

Commit:
Set up checkpointing to S3 with hadoop-aws

Commit:
Fix indentation

Commit:
Make the streaming context resumable

Resuming existing context from checkpoints if possible

Shut down gracefully with `stopGracefullyOnShutdown`

Commit:
Restructure and start to add tests

- Parameterize the S3 bucket for checkpoints
- Remove sbt-assembly plugin
- Wire up spark-packages plugin
- Bring in scalatest and spark-testing-base
- Set up travis.yml for tests, cache Ivy deps
- Remove Postgres jar from the repo (it's no longer needed with `sbt assembly`)

Commit:
Refactor Avro fixture loading in test

Pull the Avro files into the JAR as a resource, and load them via a
fixture class.

Commit:
Specify SBT version specifically in build.properties

Commit:
Add a basic unit test for counting batches

- Refactor the batch counter logic to live in functions
- Extract an `Impression` case class
- Lock ScalaTest to 2.2.1 for compatibility

Commit:
Extract a base class for test cases

http://www.scalatest.org/user_guide/defining_base_classes

Commit:
Run Scalastyle with tests

Commit:
Change `viewer_id` to an Option[String]

Commit:
Implement basic Redis persistence and tests

- Move Impression case class into its own file
- Extract objects for Post/Author counting
- Build a Redis fixture trait for testing

Commit:
Use spark-redis’s method for URI parsing

Commit:
WIP Redis/PG-based counts loader

Commit:
Tidy up a few scalastyle warnings

Commit:
Move to flat file loads/checkpoints

- Get rid of PG integration
- Rename classes
- Change project name back to `ello-spark`
- Wire up initial state loading
- Move all stream initialization into the creation function so we can actually resume from a checkpoint

Commit:
Tweak configuration for flat file checkpoints

Nicely handle cold-start scenarios without a seed file.

Coalesce to a single file to make the S3 writes more performant. May need to reevaluate with a larger data set and/or on EMR.

Add logging too.

Commit:
Get rid of a merge artifact

Commit:
Make sure to use container-based travis builds

Commit:
Split Spark batch interval and KCL checkpoint interval

Make them independently-controllable via args.

Commit:
Re-enable the streaming WAL

Commit:
Disable verbose logging
